Nil groups make panic in MatchData

- Your operating system: `sw_vers`
  - ProductName:    Mac OS X
  - ProductVersion: 10.13.1
  - BuildVersion:   17B48
- How did you install Goby: `go get github.com/goby-lang/goby`
- Goby version: 0.1.3
- The value of `$GOBY_ROOT`: /Users/kachick/gocode/src/github.com/goby-lang/goby

```ruby
'b'.match(Regexp.new('(a)?'))
```

```
panic: runtime error: index out of range
```

- Error messages

```
# goby -i
Goby 0.1.3 👺 🤠 😎
» 'b'.match(Regexp.new('(a)?'))
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/goby-lang/goby/vendor/github.com/dlclark/regexp2.(*Capture).String(...)
        /Users/kachick/gocode/src/github.com/goby-lang/goby/vm/match_data.go:123
github.com/goby-lang/goby/vm.(*VM).initMatchDataObject(0xc4200b42c0, 0xc4201da000, 0xc42019c094, 0x4, 0xc42019c048, 0x1, 0x0)
        /Users/kachick/gocode/src/github.com/goby-lang/goby/vm/match_data.go:123 +0x506
github.com/goby-lang/goby/vm.builtinStringInstanceMethods.func28.1(0xc4201360f0, 0xc42019e390, 0x1, 0x1, 0x0, 0xc42019e390, 0x0)
        /Users/kachick/gocode/src/github.com/goby-lang/goby/vm/string.go:1063 +0x107
github.com/goby-lang/goby/vm.(*thread).evalCallFrame(0xc4201360f0, 0x480d1c0, 0xc4201c2360)
        /Users/kachick/gocode/src/github.com/goby-lang/goby/vm/thread.go:63 +0x214
github.com/goby-lang/goby/vm.(*thread).startFromTopFrame(0xc4201360f0)
        /Users/kachick/gocode/src/github.com/goby-lang/goby/vm/thread.go:42 +0x61
github.com/goby-lang/goby/vm.(*thread).evalBuiltinMethod(0xc4201360f0, 0x480c8a0, 0xc4201c2200, 0xc4200ef980, 0x0, 0x1, 0xc4201be3c0, 0x0, 0x1, 0x4565223, ...)
        /Users/kachick/gocode/src/github.com/goby-lang/goby/vm/thread.go:240 +0x3f8
github.com/goby-lang/goby/vm.glob..func25(0xc4201360f0, 0x1, 0xc4201c21e0, 0xc42011c540, 0x4, 0x4)
        /Users/kachick/gocode/src/github.com/goby-lang/goby/vm/instruction.go:487 +0x4d1
github.com/goby-lang/goby/vm.(*thread).execInstruction(0xc4201360f0, 0xc4201c21e0, 0xc42011c580)
        /Users/kachick/gocode/src/github.com/goby-lang/goby/vm/thread.go:124 +0x70
github.com/goby-lang/goby/vm.(*thread).evalCallFrame(0xc4201360f0, 0x480d260, 0xc4201c21e0)
        /Users/kachick/gocode/src/github.com/goby-lang/goby/vm/thread.go:50 +0x38f
github.com/goby-lang/goby/vm.(*thread).startFromTopFrame(0xc4201360f0)
        /Users/kachick/gocode/src/github.com/goby-lang/goby/vm/thread.go:42 +0x61
github.com/goby-lang/goby/vm.(*VM).startFromTopFrame(0xc4200b42c0)
        /Users/kachick/gocode/src/github.com/goby-lang/goby/vm/vm.go:269 +0x2f
github.com/goby-lang/goby/vm.(*VM).REPLExec(0xc4200b42c0, 0xc4201c6008, 0x1, 0x1)
        /Users/kachick/gocode/src/github.com/goby-lang/goby/vm/repl.go:43 +0x78f
github.com/goby-lang/goby/igb.StartIgb(0x45654fb, 0x5)
        /Users/kachick/gocode/src/github.com/goby-lang/goby/igb/repl.go:230 +0xd23
main.main()
        /Users/kachick/gocode/src/github.com/goby-lang/goby/goby.go:29 +0x954
```

--- a/vm/match_data.go
+++ b/vm/match_data.go
@@ -3,7 +3,6 @@ package vm
 import (
 	"encoding/json"
 	"fmt"
-	"reflect"
 
 	"github.com/dlclark/regexp2"
 	"github.com/goby-lang/goby/vm/classes"
@@ -17,12 +16,10 @@ import (
 // ```
 //
 // - `MatchData.new` is not supported.
+type Match = regexp2.Match
 type MatchDataObject struct {
 	*baseObj
-	captures  []string
-	positions []int
-	pattern   string // original regex
-	text      string // original text
+	match *Match
 }
 
 // Class methods --------------------------------------------------------
@@ -58,14 +55,15 @@ func builtinMatchDataInstanceMethods() []*BuiltinMethodObject {
 					if len(args) != 0 {
 						return t.vm.initErrorObject(errors.ArgumentError, sourceLine, "Expect 0 argument. got=%d", len(args))
 					}
+					offset := 1
+					g := receiver.(*MatchDataObject).match.Groups()
+					n := len(g) - offset
+					destCaptures := make([]Object, n, n)
 
-					matchData, _ := receiver.(*MatchDataObject)
-
-					sourceCaptures := matchData.captures[1:]
-					destCaptures := make([]Object, len(sourceCaptures), len(sourceCaptures))
-
-					for i, capture := range sourceCaptures {
-						destCaptures[i] = t.vm.initStringObject(capture)
+					for _, c := range g {
+						if c.Index != 0 {
+							destCaptures[c.Index-offset] = t.vm.initStringObject(c.String())
+						}
 					}
 
 					return t.vm.initArrayObject(destCaptures)
@@ -90,12 +88,12 @@ func builtinMatchDataInstanceMethods() []*BuiltinMethodObject {
 						return t.vm.initErrorObject(errors.ArgumentError, sourceLine, "Expect 0 argument. got=%d", len(args))
 					}
 
-					matchData, _ := receiver.(*MatchDataObject)
+					g := receiver.(*MatchDataObject).match.Groups()
+					n := len(g)
+					destCaptures := make([]Object, n, n)
 
-					destCaptures := make([]Object, len(matchData.captures), len(matchData.captures))
-
-					for i, capture := range matchData.captures {
-						destCaptures[i] = t.vm.initStringObject(capture)
+					for _, c := range g {
+						destCaptures[c.Index] = t.vm.initStringObject(c.String())
 					}
 
 					return t.vm.initArrayObject(destCaptures)
@@ -132,24 +130,10 @@ func builtinMatchDataInstanceMethods() []*BuiltinMethodObject {
 // Initializes a MatchDataObject from a Match object, and the original pattern/text.
 // Nothing prevents the programmer to pass pattern/text unrelated to the match, but this will
 // create an inconsistent MatchData object.
-func (vm *VM) initMatchDataObject(match *regexp2.Match, pattern, text string) *MatchDataObject {
-	captures := make([]string, len(match.Groups()), len(match.Groups()))
-	positions := make([]int, len(match.Groups()), len(match.Groups()))
-
-	for i, group := range match.Groups() {
-		// Using as reference the Ruby MatchData implementation, we have a crucial difference with this
-		// Go implementation; the former stores only the first capture when matching, while the latter
-		// stores all, and uses as reference the last.
-		captures[i] = group.Captures[0].String()
-		positions[i] = group.Captures[0].Index
-	}
-
+func (vm *VM) initMatchDataObject(match *Match, pattern, text string) *MatchDataObject {
 	return &MatchDataObject{
-		baseObj:   &baseObj{class: vm.topLevelClass(classes.MatchDataClass)},
-		captures:  captures,
-		positions: positions,
-		pattern:   pattern,
-		text:      text,
+		baseObj: &baseObj{class: vm.topLevelClass(classes.MatchDataClass)},
+		match:   match,
 	}
 }
 
@@ -170,12 +154,11 @@ func (m *MatchDataObject) Value() interface{} {
 // returns a string representation of the object
 func (m *MatchDataObject) toString() string {
 	result := "#<MatchData"
+	result += fmt.Sprintf(" \"%s\"", m.match.Capture.String())
 
-	for i, capture := range m.captures {
-		if i == 0 {
-			result += fmt.Sprintf(" \"%s\"", capture)
-		} else {
-			result += fmt.Sprintf(" %d:\"%s\"", i, capture)
+	for _, c := range m.match.Groups() {
+		if c.Index != 0 {
+			result += fmt.Sprintf(" %d:\"%s\"", c.Index, c.String())
 		}
 	}
 
@@ -188,8 +171,8 @@ func (m *MatchDataObject) toString() string {
 func (m *MatchDataObject) toJSON() string {
 	capturesMap := make(map[int]string)
 
-	for i, capture := range m.captures {
-		capturesMap[i] = capture
+	for _, c := range m.match.Groups() {
+		capturesMap[c.Index] = c.String()
 	}
 
 	capturesJson, _ := json.Marshal(capturesMap)
@@ -199,8 +182,5 @@ func (m *MatchDataObject) toJSON() string {
 
 // equal checks if the string values between receiver and argument are equal
 func (m *MatchDataObject) equal(other *MatchDataObject) bool {
-	return reflect.DeepEqual(m.captures, other.captures) &&
-		reflect.DeepEqual(m.positions, other.positions) &&
-		m.pattern == other.pattern &&
-		m.text == other.text
+	return m.match == other.match
 }

--- a/vm/match_data.go
+++ b/vm/match_data.go
@@ -1,7 +1,6 @@
 package vm
 
 import (
-	"encoding/json"
 	"fmt"
 
 	"github.com/dlclark/regexp2"
@@ -56,14 +55,13 @@ func builtinMatchDataInstanceMethods() []*BuiltinMethodObject {
 						return t.vm.initErrorObject(errors.ArgumentError, sourceLine, "Expect 0 argument. got=%d", len(args))
 					}
 					offset := 1
-					g := receiver.(*MatchDataObject).match.Groups()
-					n := len(g) - offset
+
+					g := receiver.(*MatchDataObject).match
+					n := len(g.Groups()) - offset
 					destCaptures := make([]Object, n, n)
 
-					for _, c := range g {
-						if c.Index != 0 {
-							destCaptures[c.Index-offset] = t.vm.initStringObject(c.String())
-						}
+					for i := 0; i < n; i++ {
+						destCaptures[i] = t.vm.initStringObject(g.GroupByNumber(i + offset).String())
 					}
 
 					return t.vm.initArrayObject(destCaptures)
@@ -88,12 +86,12 @@ func builtinMatchDataInstanceMethods() []*BuiltinMethodObject {
 						return t.vm.initErrorObject(errors.ArgumentError, sourceLine, "Expect 0 argument. got=%d", len(args))
 					}
 
-					g := receiver.(*MatchDataObject).match.Groups()
-					n := len(g)
+					g := receiver.(*MatchDataObject).match
+					n := len(g.Groups())
 					destCaptures := make([]Object, n, n)
 
-					for _, c := range g {
-						destCaptures[c.Index] = t.vm.initStringObject(c.String())
+					for i := 0; i < n; i++ {
+						destCaptures[i] = t.vm.initStringObject(g.GroupByNumber(i).String())
 					}
 
 					return t.vm.initArrayObject(destCaptures)
@@ -154,12 +152,9 @@ func (m *MatchDataObject) Value() interface{} {
 // returns a string representation of the object
 func (m *MatchDataObject) toString() string {
 	result := "#<MatchData"
-	result += fmt.Sprintf(" \"%s\"", m.match.Capture.String())
 
 	for _, c := range m.match.Groups() {
-		if c.Index != 0 {
-			result += fmt.Sprintf(" %d:\"%s\"", c.Index, c.String())
-		}
+		result += fmt.Sprintf(" %s:\"%s\"", c.Name, c.String())
 	}
 
 	result += ">"
@@ -169,15 +164,15 @@ func (m *MatchDataObject) toString() string {
 
 // returns a `{ captureNumber: captureValue }` JSON-encoded string
 func (m *MatchDataObject) toJSON() string {
-	capturesMap := make(map[int]string)
+	result := "{"
 
 	for _, c := range m.match.Groups() {
-		capturesMap[c.Index] = c.String()
+		result += fmt.Sprintf(" %s:\"%s\"", c.Name, c.String())
 	}
 
-	capturesJson, _ := json.Marshal(capturesMap)
+	result += "}"
 
-	return string(capturesJson)
+	return result
 }
 
 // equal checks if the string values between receiver and argument are equal

--- a/vm/match_data.go
+++ b/vm/match_data.go
@@ -112,9 +112,9 @@ func builtinMatchDataInstanceMethods() []*BuiltinMethodObject {
 						return t.vm.initErrorObject(errors.ArgumentError, sourceLine, "Expect 0 argument. got=%d", len(args))
 					}
 
-					matchData, _ := receiver.(*MatchDataObject)
+					m := receiver.(*MatchDataObject).match
 
-					return t.vm.initIntegerObject(len(matchData.captures))
+					return t.vm.initIntegerObject(m.GroupCount())
 				}
 			},
 		},

--- a/vm/match_data_test.go
+++ b/vm/match_data_test.go
@@ -27,9 +27,9 @@ func TestMatchDataCaptures(t *testing.T) {
 		input    string
 		expected string
 	}{
-		{`'a1bca2'.match(Regexp.new('a.')).to_s`, "#<MatchData \"a1\">"},
-		{`'a1bca2'.match(Regexp.new('(a.)')).to_s`, "#<MatchData \"a1\" 1:\"a1\">"},
-		{`'a1bca2'.match(Regexp.new('(a.)(b.)')).to_s`, "#<MatchData \"a1bc\" 1:\"a1\" 2:\"bc\">"},
+		{`'a1bca2'.match(Regexp.new('a.')).to_s`, "#<MatchData 0:\"a1\">"},
+		{`'a1bca2'.match(Regexp.new('(a.)')).to_s`, "#<MatchData 0:\"a1\" 1:\"a1\">"},
+		{`'a1bca2'.match(Regexp.new('(a.)(b.)')).to_s`, "#<MatchData 0:\"a1bc\" 1:\"a1\" 2:\"bc\">"},
 	}
 
 	for i, tt := range tests {

--- a/vm/regexp.go
+++ b/vm/regexp.go
@@ -34,9 +34,10 @@ import (
 //
 // **To Goby maintainers**: avoid using Go's standard regexp package (slow and not rich). Consider the faster `Trim` or `Split` etc in Go's "strings" package first, or just use the dlclark/regexp2 instead.
 // ToDo: Regexp literals with '/.../'
+type Regexp = regexp2.Regexp
 type RegexpObject struct {
 	*baseObj
-	Regexp *regexp2.Regexp
+	regexp *Regexp
 }
 
 // Class methods --------------------------------------------------------
@@ -132,7 +133,7 @@ func builtinRegexpInstanceMethods() []*BuiltinMethodObject {
 						return t.vm.initErrorObject(errors.TypeError, sourceLine, errors.WrongArgumentTypeFormat, classes.StringClass, arg.Class().Name)
 					}
 
-					re := receiver.(*RegexpObject).Regexp
+					re := receiver.(*RegexpObject).regexp
 					m, _ := re.MatchString(input.value)
 
 					return toBooleanObject(m)
@@ -153,7 +154,7 @@ func (vm *VM) initRegexpObject(regexp string) *RegexpObject {
 	}
 	return &RegexpObject{
 		baseObj: &baseObj{class: vm.topLevelClass(classes.RegexpClass)},
-		Regexp:  r,
+		regexp:  r,
 	}
 }
 
@@ -168,12 +169,12 @@ func (vm *VM) initRegexpClass() *RClass {
 
 // Value returns the object
 func (r *RegexpObject) Value() interface{} {
-	return r.toString()
+	return r.regexp.String()
 }
 
 // toString returns the object's name as the string format
 func (r *RegexpObject) toString() string {
-	return r.Regexp.String()
+	return r.regexp.String()
 }
 
 // toJSON just delegates to toString

--- a/vm/string.go
+++ b/vm/string.go
@@ -261,7 +261,7 @@ func builtinStringInstanceMethods() []*BuiltinMethodObject {
 
 					arg := args[0]
 
-					regexp, ok := arg.(*RegexpObject)
+					re, ok := arg.(*RegexpObject)
 
 					if !ok {
 						return t.vm.initErrorObject(errors.TypeError, sourceLine, errors.WrongArgumentTypeFormat, classes.RegexpClass, arg.Class().Name)
@@ -269,7 +269,7 @@ func builtinStringInstanceMethods() []*BuiltinMethodObject {
 
 					text := receiver.(*StringObject).value
 
-					match, _ := regexp.Regexp.FindStringMatch(text)
+					match, _ := re.regexp.FindStringMatch(text)
 
 					if match == nil {
 						return NULL
@@ -1035,7 +1035,7 @@ func builtinStringInstanceMethods() []*BuiltinMethodObject {
 			// 'pow'.match(Regexp.new("x")) # => nil
 			// ```
 			//
-			// @param string [String]
+			// @param string [Regexp]
 			// @return [MatchData]
 			Name: "match",
 			Fn: func(receiver Object, sourceLine int) builtinMethodBody {
@@ -1051,16 +1051,16 @@ func builtinStringInstanceMethods() []*BuiltinMethodObject {
 						return t.vm.initErrorObject(errors.TypeError, sourceLine, errors.WrongArgumentTypeFormat, classes.RegexpClass, arg.Class().Name)
 					}
 
-					regexp := regexpObj.Regexp
+					re := regexpObj.regexp
 					text := receiver.(*StringObject).value
 
-					match, _ := regexp.FindStringMatch(text)
+					match, _ := re.FindStringMatch(text)
 
 					if match == nil {
 						return NULL
 					}
 
-					return t.vm.initMatchDataObject(match, regexp.String(), text)
+					return t.vm.initMatchDataObject(match, re.String(), text)
 				}
 			},
 		},

--- a/vm/string_test.go
+++ b/vm/string_test.go
@@ -901,7 +901,7 @@ func TestStringMatch(t *testing.T) {
 		input    string
 		expected interface{}
 	}{
-		{`"Goby!!".match(Regexp.new("G(o)b(y)")).to_s`, "#<MatchData \"Goby\" 1:\"o\" 2:\"y\">"},
+		{`"Goby!!".match(Regexp.new("G(o)b(y)")).to_s`, "#<MatchData 0:\"Goby\" 1:\"o\" 2:\"y\">"},
 		{`"Ruby".match(Regexp.new("G(o)b(y)"))`, nil},
 	}
 

